lib/spack/spack/extensions.py

```diff
@@ -11,10 +11,17 @@
 import types
 
 import llnl.util.lang
-import llnl.util.tty as tty
 import spack.config
+from spack.error import SpackError
 
-extension_regexp = re.compile(r'spack-([\w]*)')
+_extension_regexp = re.compile(r'spack-(\w[-\w]*)$')
+
+
+# TODO: For consistency we should use spack.cmd.python_name(), but
+#       currently this would create a circular relationship between
+#       spack.cmd and spack.extensions.
+def _python_name(cmd_name):
+    return cmd_name.replace('-', '_')
```

Comment:
I agree the logic could be extended. However, there may be additional checks specifically related to Spack conventions for command names (no capitals maybe?) and in that case I think this is a good place for the command.

If I'm wrong about that and this would only ever check python naming conventions then I agree `llnl.util.lang` would be a good place for it.

Reply:
Seems reasonable, although I think there's a distinction between a function to _ensure_ a legal identifier name like `python_name()`, and one to _verify_ same like `require_python_name().` The former is useful for things such as identifying what source file to import based on a command name, but there is a big space of possible command names that cannot be simply converted to a python name per https://stackoverflow.com/questions/12700893/how-to-check-if-a-string-is-a-valid-python-identifier-including-keyword-check. I think it would make more sense to retain `python_name()` as a simple substitution of `-` -> `_` (and symmetric opposite of `cmd_name()`) and implement `require_cmd_name()` in terms of `require_python_name()` and `cmd_name()` to guarantee that `python_name(cmd_name)` will always produce a legal Python identifier. Following that reasoning, `require_python_name()` would more naturally belong in `llnl.util.lang`, with the rest remaining in `spack.cmd` or a related low-level module. Thoughts?